README.md

                               t  e  l  i  z  e
                           _______________________
                     ______\                     /_______
                    \\     \\           ___     //      /
               __    \ ____  \   __    /   \   _____/\ / ____
           ___/  \____/  _//____/  \___\___/___\__   /__/  _//____
          \\__    ____  __/  __     __      ____    ____  __/  __///
            /      \_   |/    \_     /       \/     /_/   |/    \_
          \\\   ___/\___       /____/\_______/\   ___/\___       /
      <0(--- \__/ -h7- \______/   \       .    \__/ ---- \______/ --(0>
                           \      .\     /.      .
                            \      .\   //      /
                             \______\\ //______/
                                      Y

## Description

Telize is a REST API built in C with Kore allowing to get a visitor IP
address and to query location information from any IP address. It outputs
JSON-encoded IP geolocation data, and supports both JSON and JSONP.

Geolocation operations are performed using libmaxminddb which caches the
database in RAM. Therefore, Telize has very minimal overhead and should
be blazing fast.

## Requirements

### Dependencies

Telize requires [Kore][1] 4.2.0+ and [libmaxminddb][2].

### GeoIP2 databases

Telize requires the free [GeoLite2 databases][3] from MaxMind.

The path to the `GeoLite2 City` and `GeoLite2 ASN` databases can be configured
in `conf/build.conf` via the `GEOIP2DIR` macro, pointing to `/var/db/GeoIP`
by default.

Databases names can be specified using the `GEOIP2DB_CITY` and `GEOIP2DB_ASN`
macros.

## Building

Using the kodev tool:

	kodev build

## Configuration

By default, Telize will only listen on `127.0.0.1` and `::1`. To add more
listeners, `bind` directives should be added in `conf/telize.conf` in the
server blocks.

## Running

Using the kodev tool:

	kodev run

## Access logs

The default Telize configuration does not have logging enabled, it must be
configured manually. There are commented out `accesslog` directives in
`conf/telize.conf` for each domain block.

If your Telize instance produces lots of logs, [Logswan][4] might be of
interest.

## Telize and proxies

Telize handles the 'X-Forwarded-For' HTTP header if present, and returns
data for the first IP address of the list.

## CORS Support (Cross-origin resource sharing)

Telize has CORS enabled by default with the following policy:

	Access-Control-Allow-Origin: *

## Timezone offsets

Since version 3.0.0, Telize now dynamically calculates timezone offsets
(UTC time offset) and adds data to the payload.

On Alpine Linux, the 'tzdata' package needs to be installed, otherwise
localtime() will not return any data.

## Usage

For complete API documentation and usage examples, please check the
project site.

### Get IP address in Plain text format

- Example: http://127.0.0.1:8080/ip

### Get IP address in JSON format

- Example (JSON): http://127.0.0.1:8080/jsonip
- Example (JSONP): http://127.0.0.1:8080/jsonip?callback=getip

### Get IP address location in JSON format

Calling the API endpoint without any parameter will return the visitor
IP address:

- Example (JSON): http://127.0.0.1:8080/location
- Example (JSONP): http://127.0.0.1:8080/location?callback=getgeoip

Appending an IP address as parameter will return location information for
the given address:

- Example (JSON): http://127.0.0.1:8080/location/46.19.37.108
- Example (JSONP): http://127.0.0.1:8080/location/46.19.37.108?callback=getgeoip

## License

Telize is released under the BSD 2-Clause license. See `LICENSE` file
for details.

## Author

Telize is developed by Frederic Cambus.

- Site: https://www.cambus.net

## Resources

Project homepage: https://www.telize.com

Latest tarball release: https://www.statdns.com/telize/telize-3.1.1.tar.gz

GitHub: https://github.com/fcambus/telize

[1]: https://kore.io
[2]: https://github.com/maxmind/libmaxminddb
[3]: https://dev.maxmind.com/geoip/geoip2/geolite2/
[4]: https://www.logswan.org


ChangeLog
Telize 3.1.1 (2022-04-27)

- Define GEOIP2DB_CITY and GEOIP2DB_ASN macros to load MMDB databases
- Account for Kore 4.2 changes in http_request structure
- Account for Kore 4.2 changes in the routing configuration



Telize 3.1.0 (2021-03-29)

- Decrement COUNTRIES to account for "Netherlands Antilles" removal
- Allow telize_getdata() to handle boolean values from MMDB databases
- Add the 'is_in_european_union' field if it's present in the MMDB database



Telize 3.0.1 (2021-01-01)

- Add port number for endpoints examples, so they work with default settings
- Add a 'Timezone offsets' section in the README
- Remove entry for "Netherlands Antilles" in 'country_code' and 'country_code3'
  arrays
- Add -Wextra and -pedantic to CFLAGS
- Enable FORTIFY_SOURCE level 2
- Check strndup() return value and error out if allocation fails



Telize 3.0.0 (2020-09-07)

- Telize has been rewritten in C using the Kore Web platform
- Add a few entries in the 'country_code' to 'country_code3' map
- Handle the 'X-Forwarded-For' HTTP header if present
- When 'X-Forwarded-For' header is set, get the first IP of the list
- Dynamically calculate timezone offsets and add data to the payload



Telize 2.0.0 (2018-03-15)

- Switch to using GeoIP2/GeoLite2 databases
- The 'region_code' field is now in the ISO-3166-2 format for all countries
- The 'organization' field now only contains organization name, Autonomous
  System Numbers are now provided as a separate 'asn' field in numeric value
- Timezone offsets ('offset' field) are now generated as numeric values
  and in seconds
- The 'area_code' and 'dma_code' are not provided anymore
- Drop dependency on the Lua iconv module, GeoIP2/GeoLite2 databases are
  encoded as UTF-8
- Drop timezone.conf, GeoIP2/GeoLite2 databases contain time zones for each
  location
- Rename telize to telize.conf
- Update timezones list
- Add support for CORS in the 'jsonip' endpoint
- Rename the 'geoip' endpoint to 'location'
- Add Cache-Control header in the 'jsonip' endpoint as well



Telize 1.07 (2016-11-22)

- The timezone offset generator has been rewritten in Lua
- Relicensed under the BSD 2-Clause license
- Use Lua to set additional headers and drop dependency on the
  'ngx_headers_more' module
- Use Lua to echo IP address in the 'ip' endpoint and drop dependency on
  the 'ngx_echo' module
- Use 'content_by_lua_block' directive instead of 'content_by_lua'
- Various documentation updates



Telize 1.06 (2016-02-04)

- Prefixing all country codes with carets to prevent wrong matching



Telize 1.05 (2016-01-05)

- Regression fix: organization name was not correctly converted to UTF-8



Telize 1.04 (2015-11-10)

- Do not attempt to split GeoIP org field into ASN and ISP, return the raw
  organization string instead



Telize 1.03 (2015-10-27)

- Fix region and ISP name encoding (Thanks to Ludovic Robinot)
- Removed legacy Debian automatic install script
- Check for invalid IP addresses before attempting to validate arguments
- Convert latitude and longitude to numeric values only after all arguments
  have been validated
- The 'jsonip' endpoint do not use the internal 'jsonify' endpoint anymore
- Various documentation updates
- The 'localize' internal endpoint is not needed anymore, and Telize now
  requires Nginx 1.7.4+
- Splitting GeoIP organization string into ASN and ISP using Lua



Telize 1.02 (2014-09-01)

- Refactored Lua code (Thanks to Jon Nalley)
- Adding support for Timezones offsets (Thanks to Yann Verry)
- Default output charset is now UTF-8 (Thanks to Yann Verry)
- Adding an automatic install script for Debian
- Enable setting charsets for application/json MIME type
- Adding Cache-Control headers (no-cache by default)
- Enabling CORS by default



Telize 1.01 (2013-10-11)

- Adding support for Timezones
- Adding support for GeoIP ASN database
- Enforcing strictly defined URL endpoints
- Prevent logging of internal redirects



Telize 1.00 (2013-08-21)

- Initial release


telize.go
/*
 * Telize 3.1.0
 * Copyright (c) 2013-2022, Frederic Cambus
 * https://www.telize.com
 *
 * Created:      2013-08-15
 * Last Updated: 2022-01-18
 *
 * Telize is released under the BSD 2-Clause license.
 * See LICENSE file for details.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"github.com/go-chi/chi/v5"
	"github.com/oschwald/maxminddb-golang"
	"io"
	"log"
	"net"
	"net/http"
	"net/http/fcgi"
	"os"
	"time"
)

var asn *maxminddb.Reader
var city *maxminddb.Reader

type ASN struct {
	AutonomousSystemNumber       uint   `maxminddb:"autonomous_system_number"`
	AutonomousSystemOrganization string `maxminddb:"autonomous_system_organization"`
}

type City struct {
	Continent struct {
		Code  string            `maxminddb:"code"`
		Names map[string]string `maxminddb:"names"`
	} `maxminddb:"continent"`
	Country struct {
		IsInEuropeanUnion bool              `maxminddb:"is_in_european_union"`
		IsoCode           string            `maxminddb:"iso_code"`
		Names             map[string]string `maxminddb:"names"`
	} `maxminddb:"country"`
	Subdivisions []struct {
		IsoCode string            `maxminddb:"iso_code"`
		Names   map[string]string `maxminddb:"names"`
	} `maxminddb:"subdivisions"`
	City struct {
		Names map[string]string `maxminddb:"names"`
	} `maxminddb:"city"`
	Postal struct {
		Code string `maxminddb:"code"`
	} `maxminddb:"postal"`
	Location struct {
		Latitude  float64 `maxminddb:"latitude"`
		Longitude float64 `maxminddb:"longitude"`
		TimeZone  string  `maxminddb:"time_zone"`
	} `maxminddb:"location"`
}

type payload struct {
	IP                string  `json:"ip"`
	Continent         string  `json:"continent_code,omitempty""`
	Country           string  `json:"country,omitempty""`
	CountryCode       string  `json:"country_code,omitempty""`
	CountryCode3      string  `json:"country_code3,omitempty""`
	IsInEuropeanUnion bool    `json:"is_in_european_union,omitempty""`
	Region            string  `json:"region,omitempty""`
	RegionCode        string  `json:"region_code,omitempty""`
	City              string  `json:"city,omitempty""`
	PostalCode        string  `json:"postal_code,omitempty""`
	Latitude          float64 `json:"latitude,omitempty"`
	Longitude         float64 `json:"longitude,omitempty"`
	TimeZone          string  `json:"timezone,omitempty"`
	Offset            int     `json:"offset,omitempty"`
	ASN               uint    `json:"asn,omitempty"`
	Organization      string  `json:"organization,omitempty"`
}

func ip(w http.ResponseWriter, r *http.Request) {
	ip, _, _ := net.SplitHostPort(r.RemoteAddr)

	io.WriteString(w, ip)
}

// Generate JSON output
func jsonify(w http.ResponseWriter, r *http.Request, payload *payload) {
	callback := r.URL.Query().Get("callback")

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	if json, err := json.Marshal(payload); err == nil {
		if callback != "" {
			io.WriteString(w, callback+"("+string(json)+");")
		} else {
			io.WriteString(w, string(json))
		}
	}
}

func jsonip(w http.ResponseWriter, r *http.Request) {
	ip, _, _ := net.SplitHostPort(r.RemoteAddr)
	jsonip := payload{IP: ip}

	jsonify(w, r, &jsonip)
}

func location(w http.ResponseWriter, r *http.Request) {
	ip := chi.URLParam(r, "ip")

	ip_ := net.ParseIP(ip)

	var asn_record ASN

	err := asn.Lookup(ip_, &asn_record)
	if err != nil {
		log.Panic(err)
	}

	var record City

	err = city.Lookup(ip_, &record)
	if err != nil {
		log.Panic(err)
	}

	tz, _ := time.LoadLocation(record.Location.TimeZone)
	_, offset := time.Now().In(tz).Zone()

	jsonip := payload{
		IP:                ip,
		Continent:         record.Continent.Code,
		Country:           record.Country.Names["en"],
		CountryCode:       record.Country.IsoCode,
		CountryCode3:      CountryCode3[record.Country.IsoCode],
		IsInEuropeanUnion: record.Country.IsInEuropeanUnion,
		City:              record.City.Names["en"],
		PostalCode:        record.Postal.Code,
		Latitude:          record.Location.Latitude,
		Longitude:         record.Location.Longitude,
		TimeZone:          record.Location.TimeZone,
		Offset:            offset,
		ASN:               asn_record.AutonomousSystemNumber,
		Organization:      asn_record.AutonomousSystemOrganization,
	}

	if record.Subdivisions != nil {
		jsonip.Region = record.Subdivisions[0].Names["en"]
		jsonip.RegionCode = record.Subdivisions[0].IsoCode
	}

	jsonify(w, r, &jsonip)
}

func main() {
	var err error

	fastcgi := flag.Bool("fastcgi", false, "Enable FastCGI mode")
	host := flag.String("host", "127.0.0.1", "Set the server host")
	port := flag.String("port", "8080", "Set the server port")
	version := flag.Bool("version", false, "Display version")

	mode := "HTTP"

	flag.Usage = func() {
		fmt.Println("\nUSAGE:")
		flag.PrintDefaults()
	}
	flag.Parse()

	if *version {
		fmt.Println("Telize 4.0.0")
		os.Exit(0)
	}

	if *fastcgi {
		mode = "FastCGI"
	}

	asn, err = maxminddb.Open("GeoLite2-ASN.mmdb")
	if err != nil {
		log.Fatal(err)
	}
	defer asn.Close()

	city, err = maxminddb.Open("GeoLite2-City.mmdb")
	if err != nil {
		log.Fatal(err)
	}
	defer city.Close()

	address := *host + ":" + *port

	r := chi.NewRouter()
	r.Get("/ip", ip)
	r.Get("/jsonip", jsonip)
	r.Get("/location/{ip}", location)

	if *fastcgi {
		listener, _ := net.Listen("tcp", address)

		if err := fcgi.Serve(listener, r); err != nil {
			fmt.Println("\nERROR:", err)
			os.Exit(1)
		}
	} else {
		if err := http.ListenAndServe(address, r); err != nil {
			fmt.Println("\nERROR:", err)
			os.Exit(1)
		}
	}

	fmt.Println("Listening on ("+mode+" mode):", address)
}
